File: Java/DatabaseSQLiteManager.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

/** Una classe Java per gestire un database SQLite
  *
  */
public class DatabaseSQLiteManager {
    private String url;
    private Connection connection;

    public DatabaseSQLiteManager(String dbUrl) {
        this.url = dbUrl;
    }

    // Metodo per connettersi al database
    // Nota: SQLite crea automaticamente il file del database se non esiste già, quindi non è necessario alcun codice aggiuntivo per gestire la creazione del file
    private Connection connect() throws SQLException {
        if (connection == null || connection.isClosed()) {
            connection = DriverManager.getConnection(url);
            connection.setAutoCommit(false); // Disabilita il commit automatico
        }
        return connection;
    }

    // Metodo per eseguire una query di selezione con parametri
    public ResultSet selectData(String sql, Object... params) {
        ResultSet rs = null;
        try {
            PreparedStatement pstmt = connect().prepareStatement(sql);
            setParameters(pstmt, params); // Imposta i parametri
            rs = pstmt.executeQuery();
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        }
        // Ricorda di ciclare su rs.next() (es while (rs != null && rs.next()) { TODO }) nel chiamante
        /* Per leggere campi int rs.getInt('campo'), per leggere campi string rs.getString('campo')
        */
        return rs;
    }

    // Metodo per scrivere (inserire, aggiornare o cancellare) dati nel database
    public boolean updateData(String query, Object... params) {
        try (PreparedStatement pstmt = connect().prepareStatement(query)) {
            setParameters(pstmt, params);
            pstmt.executeUpdate();
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    // Metodo per eseguire query DDL (CREATE, ALTER, DROP, ecc... )
    public void executeQuery(String sql) {
        try (PreparedStatement pstmt = connect().prepareStatement(sql)) {
            pstmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // Metodo per impostare i parametri nel PreparedStatement
    private void setParameters(PreparedStatement pstmt, Object... params) throws SQLException {
        for (int i = 0; i < params.length; i++) {
            pstmt.setObject(i + 1, params[i]);
        }
    }

    // Metodo per ottimizzare il database
    public void optimizeDatabase() {
        String query = "VACUUM"; // Comando per ottimizzare il database
        try (PreparedStatement pstmt = connect().prepareStatement(query)) {
            pstmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // Metodo per eseguire il commit delle modifiche
    public void commitTransaction() {
        try {
            if (connection != null && !connection.isClosed()) {
                connection.commit();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // Metodo per eseguire il rollback delle modifiche
    public void rollbackTransaction() {
        try {
            if (connection != null && !connection.isClosed()) {
                connection.rollback();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // Metodo per chiudere la connessione al database
    public void closeConnection() {
        try {
            if (connection != null && !connection.isClosed()) {
                connection.close();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: PHP/DatabaseSQLiteManager.php
<?php
/**
 * Una classe PHP per gestire un database SQLite utilizzando PDO.
 *
 */
class DatabaseSQLiteManager {
    /** @var string La stringa di connessione (DSN) per SQLite. */
    private $dsn;

    /** @var PDO|null La connessione attiva al database. */
    private $pdo;

    /**
     * Costruttore.
     *
     * @param string $dbPath Il percorso al file del database SQLite (es. 'database.db').
     * Se il file non esiste, verrà creato.
     */
    public function __construct(string $dbPath) {
        // Il DSN per SQLite è 'sqlite:percorso/al/file.db'
        $this->dsn = "sqlite:" . $dbPath;
    }

    // --- Metodi di Connessione e Transazione ---

    /**
     * Stabilisce e restituisce la connessione PDO.
     * Gestisce la creazione del file del database se non esiste.
     *
     * @return PDO La connessione attiva.
     * @throws PDOException Se la connessione fallisce.
     */
    private function connect(): PDO {
        // Se la connessione non è ancora stata stabilita o è stata chiusa (in PHP si controlla null)
        if ($this->pdo === null) {
            try {
                // PDO lancia un'eccezione in caso di errore
                $this->pdo = new PDO($this->dsn);

                // Abilita la modalità per lanciare eccezioni per gli errori (importante!)
                $this->pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

                // Disabilita il commit automatico, come nell'originale Java
                $this->pdo->setAttribute(PDO::ATTR_AUTOCOMMIT, 0);

                // Avvia una transazione per emulare l'uso iniziale di setAutoCommit(false)
                $this->pdo->beginTransaction();

            } catch (PDOException $e) {
                // Equivalente a e.printStackTrace() in Java
                echo "Errore di connessione al database: " . $e->getMessage() . "\n";
                // Rilancia l'eccezione o restituisce null, a seconda della gestione degli errori desiderata
                throw $e;
            }
        }
        return $this->pdo;
    }

    /**
     * Chiude la connessione al database.
     * La chiusura esplicita in PHP non è sempre strettamente necessaria (garbage collector),
     * ma è una buona pratica per rilasciare le risorse.
     */
    public function closeConnection(): void {
        // Impostare a null chiude la connessione PDO
        $this->pdo = null;
    }

    /**
     * Esegue il commit delle modifiche (commit transaction).
     */
    public function commitTransaction(): void {
        if ($this->pdo !== null) {
            try {
                $this->pdo->commit();
                // Riavvia la transazione per continuare a disabilitare l'autocommit
                $this->pdo->beginTransaction();
            } catch (PDOException $e) {
                echo "Errore durante il commit: " . $e->getMessage() . "\n";
            }
        }
    }

    /**
     * Esegue il rollback delle modifiche (rollback transaction).
     */
    public function rollbackTransaction(): void {
        if ($this->pdo !== null) {
            try {
                $this->pdo->rollBack();
                // Riavvia la transazione
                $this->pdo->beginTransaction();
            } catch (PDOException $e) {
                echo "Errore durante il rollback: " . $e->getMessage() . "\n";
            }
        }
    }

    // --- Metodi di Esecuzione Query ---

    /**
     * Metodo per eseguire una query di selezione con parametri.
     *
     * @param string $sql La query SQL.
     * @param array $params I parametri da associare (bind).
     * @return array|null Un array di righe risultanti (ogni riga è un array associativo) o null in caso di errore.
     */
    public function selectData(string $sql, array $params = []): ?array {
        try {
            $stmt = $this->connect()->prepare($sql);

            // PDO gestisce l'associazione dei tipi in base al valore
            foreach ($params as $key => $value) {
                // PDO binding è 1-based, come in Java, ma usiamo l'indice del parametro nell'array (key + 1)
                $stmt->bindValue($key + 1, $value);
            }

            $stmt->execute();
            // In PHP, si restituisce un array completo di risultati, non un oggetto "cursore" come ResultSet
            return $stmt->fetchAll(PDO::FETCH_ASSOC);

        } catch (PDOException $e) {
            echo "Errore Select: " . $e->getMessage() . "\n";
            return null;
        }
    }

    /**
     * Metodo per scrivere (inserire, aggiornare o cancellare) dati nel database.
     *
     * @param string $query La query SQL (INSERT, UPDATE, DELETE).
     * @param array $params I parametri da associare (bind).
     * @return int Il numero di righe modificate (0 in caso di errore, o il numero di righe).
     */
    public function updateData(string $query, array $params = []): int {
        try {
            $stmt = $this->connect()->prepare($query);

            foreach ($params as $key => $value) {
                $stmt->bindValue($key + 1, $value);
            }

            $stmt->execute();
            return $stmt->rowCount(); // Restituisce il numero di righe interessate
        } catch (PDOException $e) {
            echo "Errore Update/Delete: " . $e->getMessage() . "\n";
            return 0; // 0 righe modificate
        }
    }

    /**
     * Metodo per eseguire query DDL (CREATE, ALTER, DROP, ecc...) senza parametri.
     *
     * @param string $sql La query DDL SQL.
     */
    public function executeQuery(string $sql): void {
        try {
            // Per query DDL semplici, si può usare il metodo exec di PDO
            $this->connect()->exec($sql);
        } catch (PDOException $e) {
            echo "Errore DDL: " . $e->getMessage() . "\n";
        }
    }

    /**
     * Metodo per eliminare una tabella.
     *
     * @param string $tableName Il nome della tabella da eliminare.
     */
    public function dropTable(string $tableName): void {
        $dropTableSQL = "DROP TABLE IF EXISTS " . $tableName;
        $this->executeQuery($dropTableSQL); // Riutilizza il metodo executeQuery
    }

    /**
     * Metodo per ottimizzare il database.
     */
    public function optimizeDatabase(): void {
        $this->executeQuery("VACUUM"); // Riutilizza il metodo executeQuery
    }

    // --- Metodi di utilità (Adattamento di tableExists/columnExists) ---

    /**
     * Metodo per controllare se esiste una tabella.
     *
     * @param string $tableName Il nome della tabella.
     * @return bool True se la tabella esiste, altrimenti false.
     */
    public function tableExists(string $tableName): bool {
        // Uso del metodo selectData, come il selectData in Java
        $query = "SELECT name FROM sqlite_master WHERE type = 'table' AND name = ?";
        $result = $this->selectData($query, [$tableName]);

        // Se l'array di risultati non è vuoto, la tabella esiste
        return !empty($result);
    }

    /**
     * Metodo per controllare se esiste una colonna in una tabella.
     * Nota: in PDO/PHP, questo richiede la connessione per il PRAGMA.
     *
     * @param string $tableName Il nome della tabella.
     * @param string $columnName Il nome della colonna.
     * @return bool True se la colonna esiste, altrimenti false.
     */
    public function columnExists(string $tableName, string $columnName): bool {
        try {
            // PRAGMA table_info('tableName') è specifico di SQLite
            $query = "PRAGMA table_info(" . $tableName . ")";
            $stmt = $this->connect()->query($query); // Usiamo query() per una select semplice senza parametri

            while ($row = $stmt->fetch(PDO::FETCH_ASSOC)) {
                // Il nome della colonna in PRAGMA table_info è nel campo 'name'
                if (strcasecmp($row['name'], $columnName) === 0) {
                    return true;
                }
            }
        } catch (PDOException $e) {
            echo "Errore nella verifica della colonna: " . $e->getMessage() . "\n";
        }

        return false;
    }
}

?>
